src/storage/src/table/state_table.rs

```diff
@@ -102,28 +112,124 @@ impl<S: StateStore> StateTable<S> {
-        _keyspace: Keyspace<S>,
-        _table_descs: Vec<ColumnDesc>,
-        _epoch: u64,
-        _stats: Arc<StateStoreMetrics>,
-    ) -> StorageResult<Self> {
-        todo!()
+        keyspace: Keyspace<S>,
+        table_descs: Vec<ColumnDesc>,
+        mem_table_vec: Arc<Vec<MemTableItem>>,
+        order_types: Vec<OrderType>,
+        epoch: u64,
+        stats: Arc<StateStoreMetrics>,
+    ) -> StorageResult<StateTableRowIter<S>> {
+        let cell_based_row_iter =
+            CellBasedTableRowIter::new(keyspace.clone(), table_descs, epoch, stats).await?;
+        let iter = Self {
+            cell_based_row_iter,
+            mem_table_vec,
+            mem_table_idx: 0,
+            order_types,
+            keyspace,
+        };
+
+        Ok(iter)
     }
 
+    fn mem_table_current_item(&self) -> Option<&MemTableItem> {
+        self.mem_table_vec.get(self.mem_table_idx)
+    }
+
+}
+
+#[async_trait::async_trait]
+impl<S: StateStore> TableIter for StateTableRowIter<S> {
     async fn next(&mut self) -> StorageResult<Option<Row>> {
-        todo!()
+        let cell_based_item = self.cell_based_row_iter.next_with_pk().await?;
+        let mem_table_cur_item = self.mem_table_vec.get(self.mem_table_idx);
+        match (cell_based_item, mem_table_cur_item) {
+            (None, None) => {
+                return Ok(None);
+            }
+            (Some((_, row)), None) => {
+                return Ok(Some(row));
+            }
+            (None, Some((_, row_op))) => {
+                self.mem_table_idx += 1;
+                match row_op {
+                    RowOp::Insert(row) => return Ok(Some(row.clone())),
+                    RowOp::Delete(_) => {
+                        return Ok(None);
+                    }
+                    RowOp::Update((_, new_row)) => return Ok(Some(new_row.clone())),
+                }
+            }
+            (Some((cell_based_pk, cell_based_row)), Some((mem_table_pk, mem_table_row_op))) => {
+                let pk_serializer = OrderedRowSerializer::new(self.order_types.clone());
+                let mem_table_pk_bytes =
+                    serialize_pk(mem_table_pk, &pk_serializer).map_err(err)?;
+                let mem_table_pk_with_prefix = self.keyspace.prefixed_key(mem_table_pk_bytes);
+                match cell_based_pk.cmp(&mem_table_pk_with_prefix) {
+                    Ordering::Less => {
+                        return Ok(Some(cell_based_row));
+                    }
+                    Ordering::Greater => {
+                        self.mem_table_idx += 1;
+                        match mem_table_row_op {
+                            RowOp::Insert(row) => return Ok(Some(row.clone())),
+                            RowOp::Delete(_) => {
+                                return Ok(None);
+                            }
+                            RowOp::Update((_, new_row)) => return Ok(Some(new_row.clone())),
+                        }
+                    }
```

Comment:
Seems the row from the cell-based table is dropped here? Will we lose data in this way? We need a way to peek the cell-based row iterator.

Reply:
If we use `Stream` for the iterators, then we can simply use `Peekable` for this situation. 🤣